.travis.yml

```diff
@@ -1,31 +1,65 @@
+env:
+  global:
+    # include $HOME/.local/bin for `aws`
+    - PATH=$HOME/.local/bin:$PATH
 language: shell
 notifications:
   email:
     on_success: never
 stages:
-- tests
+- unit_tests
+- build
+- e2e_tests
 - deployment
+before_install:
+  # set up awscli packages
+  - pip install --user awscli
+  # set up shared storage for build artifacts
+  - mkdir -p ~/build-stages-shared-storage/$TRAVIS_BUILD_NUMBER
+  - aws s3 sync s3://aws-log-forwader-travis-build-shared-storage/$TRAVIS_BUILD_NUMBER ~/build-stages-shared-storage/$TRAVIS_BUILD_NUMBER
 jobs:
   include:
-    - stage: tests
-      name: Tests & Linting
+    - stage: unit_tests
+      name: Execute unit tests & linting
       language: shell
       services: docker
       script:
         - docker build -f pipeline/Dockerfile -t lambda_image .
         - docker run --rm lambda_image bash -c 'python3 -m pylint --rcfile=pipeline/pylint.cfg $(find src/ -name "*.py")'
         - docker run --rm lambda_image bash -c 'cd src && python3 -m pytest -v ../tests/unit'
+    - stage: build
+      name: Build AWS Lambda package
+      #if: tag =~ /^release.*$/
+      language: shell
+      services: docker
+      script:
+        - docker run --rm -v `pwd`:/var/task amazon/aws-sam-cli-build-image-python3.8 bash -c './build-release-package.sh'
+        - cp -p dynatrace-aws-log-forwarder.zip ~/build-stages-shared-storage/$TRAVIS_BUILD_NUMBER
+        - cp -p dynatrace-aws-log-forwarder-lambda.zip ~/build-stages-shared-storage/$TRAVIS_BUILD_NUMBER
+    - stage: e2e_tests
+      name: Execute e2e tests
+      if: branch = master OR type = pull_request
+      language: shell
+      services: docker
+      script:
+        # deploy log forwarder resources reqiured for e2e to AWS
+        - cp -r ~/build-stages-shared-storage/$TRAVIS_BUILD_NUMBER/dynatrace-aws-log-forwarder-lambda.zip .
+        - ./dynatrace-aws-logs.sh deploy --target-url $DYNATRACE_ENV_URL --target-api-token $DYNATRACE_API_KEY --require-valid-certificate $VERIFY_SSL --stack-name aws-log-forwarder-e2e-build-$TRAVIS_BUILD_NUMBER

```

Comment:
Please confirm if I understand correctly. 
You are using script from repository? But with lambda.zip from previous step. If it is so, it does not test the full release zip package.

I would expect: you take release zip from build step. You copy it to new dir. Unzip in new dir. Perform deployment in there so we are sure you are not taking things from repo (but from zip as expected).
The artifact that we release it this full release zip and e2e should only be using this artifact and nothing from other places

Reply:
I'm not using the script from the repo. All the archives/packages used are products of the current build.

1. I need to copy the entire release package from the build step to the shared storage so that it is available later for the deploy stage.
2. For e2e I just need the Lambda package from the build step, as both the script and the CF template are parts of the repo. To zip all parts, copy them and unzip them in the next step would be overengineering and cause trouble (overwriting repo files)

Reply:
Let's say someone does some change to build-release-package.sh and in result one of files is not included in the release zip. The test would not find this.
E2E test should only use release resources, not touch anything from repo to. To avoid overwrite you can do this unzip in new temp folder. 
E2E idea is not to deploy the thing as easily as possible but to deploy the thing using 100% same resources as client is using, in this case only release.zip

Reply:
`To zip all parts, copy them and unzip them in the next step would be overengineering and cause trouble (overwriting repo files)` you don't have to zip them and copy, you already have the artifact from build step

Reply:
Updated the pipeline, so that e2e only utilizes build package content.